tfhe/src/core_crypto/gpu/slice.rs

```diff
@@ -0,0 +1,194 @@
+use crate::core_crypto::gpu::vec::range_bounds_to_start_end;
+use crate::core_crypto::gpu::{CudaPtr, CudaStream};
+use crate::core_crypto::prelude::Numeric;
+use std::ffi::c_void;
+use std::marker::PhantomData;
+use tfhe_cuda_backend::cuda_bind::{cuda_memcpy_async_gpu_to_gpu, cuda_memcpy_async_to_cpu};
+
+#[derive(Debug)]
+pub struct CudaSlice<'a, T: Numeric> {
+    ptr: CudaPtr,
+    _len: usize,
+    _phantom_1: PhantomData<T>,
+    _phantom_2: PhantomData<&'a ()>,
+}
+
+#[derive(Debug)]
+pub struct CudaSliceMut<'a, T: Numeric> {
+    ptr: CudaPtr,
+    len: usize,
+    _phantom_1: PhantomData<T>,
+    _phantom_2: PhantomData<&'a mut ()>,
+}
+
+impl<'a, T> CudaSlice<'a, T>
+where
+    T: Numeric,
+{
+    /// # Safety
+    pub(crate) unsafe fn new(ptr: CudaPtr, len: usize) -> CudaSlice<'a, T> {
+        CudaSlice {
+            ptr,
+            _len: len,
+            _phantom_1: PhantomData,
+            _phantom_2: PhantomData,
+        }
+    }
+
+    /// # Safety
+    pub(crate) unsafe fn as_c_ptr(&self) -> *const c_void {
+        self.ptr.as_c_ptr()
+    }
+}
+
+impl<'a, T> CudaSliceMut<'a, T>
+where
+    T: Numeric,
+{
+    pub(crate) unsafe fn new(ptr: CudaPtr, len: usize) -> CudaSliceMut<'a, T> {
+        CudaSliceMut {
+            ptr,
+            len,
+            _phantom_1: PhantomData,
+            _phantom_2: PhantomData,
+        }
+    }
+
+    /// # Safety
+    pub(crate) unsafe fn as_mut_c_ptr(&mut self) -> *mut c_void {
+        self.ptr.as_mut_c_ptr()
+    }
+
+    /// # Safety
+    pub(crate) unsafe fn as_c_ptr(&self) -> *const c_void {
+        self.ptr.as_c_ptr()
+    }
+
+    /// Copies data between two `CudaSlice`
+    ///
+    /// # Safety
+    ///
+    /// - [CudaStream::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_from_gpu_async(&mut self, src: &Self, stream: &CudaStream)
+    where
+        T: Numeric,
+    {
+        assert!(self.len() >= src.len());
+        let size = src.len() * std::mem::size_of::<T>();
+        // We check that src is not empty to avoid invalid pointers
+        if size > 0 {
+            cuda_memcpy_async_gpu_to_gpu(
+                self.as_mut_c_ptr(),
+                src.as_c_ptr(),
+                size as u64,
+                stream.as_c_ptr(),
+            );
+        }
+    }
+
+    #[allow(dead_code)]
+    /// Copies data of a `CudaSlice` to the CPU
+    ///
+    /// # Safety
+    ///
+    /// - [CudaStream::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_to_cpu_async(&self, dest: &mut [T], stream: &CudaStream)
+    where
+        T: Numeric,
+    {
+        assert!(self.len() <= dest.len());
```

Comment:
I think we shold test for equality, thus making the user have explicit and correct range for both slices